### PR TITLE
sparrow: init at 1.6.4

### PR DESCRIPTION
**Pull request by emmanuelrosa**
###### Description of changes

This change adds a package for Sparrow, a Bitcoin wallet.

I've had this package in a Nix Flake for about 4 months, and the application has been working. There were two remaining issues which were recently fixed:

 - QR code scanning did not work. `patchelf` didn't work with OpenIMAJWrapper.so, so I now compile only that library.
 - Tor connectivity using the built-in Tor. However, keep in mind that Sparrow hard-codes the Tor proxy port 9050. So if you have a Tor proxy already running on that port, the you'll need to use that proxy instead of using the built-in Tor.

This change also includes (generic) helper functions to generate desktop icons; Java applications don't usually install icons. They work similar to makeDesktopItem, and make a good companion. Let me know if these functions would be useful for other packages, so that I know to submit a future PR to make them available. In fact, I'd like to update my Bisq package to use them.

Note that with certain window managers, such as bspwm, there are issues when clicking on the menu items. This is a JDK/JFX issue. See https://github.com/sparrowwallet/sparrow/issues/170

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emmanuelrosa** [comment]
Thank you @_1000101 for your interest in creating this package. I hope you don't mind being a maintainer with me.

**Comment by 1000101** [comment]
Result of `nixpkgs-review pr 167727` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>sparrow</li>
  </ul>
</details>

**Comment by emmanuelrosa** [comment]
I added udev rules for hardware wallets.

**Comment by emmanuelrosa** [comment]
There's an issue with the bundled `hwi` which Sparrow uses to communicate with wired hardware wallets. The problem is it's a self-extracting executable of the original Python `hwi`. and it's unable to extract itself. The archive which comes with the executable is appended to the tail of a "bootloader" when the executable is constructed; It's built with PyInstaller. I checked the tail of the file, and found references to paths in the Nix store. This suggests that patchelf is overriding part of the archive, thus corrupting it.

Ideally, we'd use the `hwi` Nix package instead, however that package is currently broken because it depends on an old version of a BitBox2 Python package.

I think the ideal solution is to wait on the next release of `hwi`, since it uses the latest BitBox2 package. Right now, the next release of `hwi` is a release candidate, so perhaps it will be ready in the near future. So let's wait on that before merging this.

**Comment by prusnak** [comment]
> I think the ideal solution is to wait on the next release of `hwi`, since it uses the latest BitBox2 package.

HWI has been updated 2 weeks ago in https://github.com/NixOS/nixpkgs/pull/168836

Can you give using HWI from nixpkgs another shot?

**Comment by emmanuelrosa** [comment]
The HWI executable now runs. I'm using the one from Nixpkgs.

File: pkgs/applications/blockchains/sparrow/openimajgrabber.nix
Comment (prusnak) [inline pkgs/applications/blockchains/sparrow/openimajgrabber.nix:10]: Please use `fetchFromGitHub`

File: pkgs/applications/blockchains/sparrow/openimajgrabber.nix
Comment (prusnak) [inline pkgs/applications/blockchains/sparrow/openimajgrabber.nix:35]: Since we are building from source we can use `platforms = platforms.linux;`

File: pkgs/applications/blockchains/sparrow/openimajgrabber.nix
Comment (prusnak) [inline pkgs/applications/blockchains/sparrow/openimajgrabber.nix:19]: Please add a comment that these build instructions come from `build.sh`

**Review by prusnak — CHANGES_REQUESTED** [review]
Let some nits in `openimajgrabber` package.

Couple of other things:
* please use `init at` instead of `init @` in the commit message
* is it really necessary to introduce `copyDesktopIcons` and `makeDesktopIcon` - seems like a lot of unnecessary scaffolding for such small functionality

**Comment by prusnak** [comment]
Can you please update this to Sparrow 1.6.4 please? I'll review immediately afterwards.

File: pkgs/applications/blockchains/sparrow/default.nix
Comment (0xB10C) [inline pkgs/applications/blockchains/sparrow/default.nix:189]: This fails for me.

```
building '/nix/store/i4abinzh1sdpk0yxj9q5ijmsrw0c0lxn-Sparrow.desktop.drv'...
building '/nix/store/g7viam47w8zkyy73ga406pib2q2hfpiq-dbus-catalog.xml.drv'...
Running desktop-file validation
/nix/store/jcjsmg8d7rj64lv7bnz0gnrl43pvx9qq-Sparrow.desktop/share/applications/Sparrow.desktop: error: value "Office Finance" for key "Categories" in group "Desktop Entry" contains an unregistered value "Office Finance"; values extending the format should start with "X-"
/nix/store/jcjsmg8d7rj64lv7bnz0gnrl43pvx9qq-Sparrow.desktop/share/applications/Sparrow.desktop: hint: value "Office Finance" for key "Categories" in group "Desktop Entry" does not contain a registered main category; application might only show up in a "catch-all" section of the application menu
builder for '/nix/store/i4abinzh1sdpk0yxj9q5ijmsrw0c0lxn-Sparrow.desktop.drv' failed with exit code 1
```

I guess this could be just:
```suggestion
      categories = ["Finance" ];
```
 [`Finance` seems to belong to `Office`](https://specifications.freedesktop.org/menu-spec/latest/apas02.html)

**Review by 0xB10C — CHANGES_REQUESTED** [review]

**Pull request merged**
